In [1]:
import pandas as pd
import numpy as np 

In [2]:
data_1 = pd.read_csv(r"C:\Users\DELL\Desktop\Scaler Numpy and Pandas\gapminder.csv")

In [3]:
data_1

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1699     Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298

[1704 rows x 6 columns]

# Creating a DataFrame. (Two methods - Row wise and Column wise).

# Row-wise

In [6]:
df1 = pd.DataFrame([[10,20,30]] , columns = ["A" , "B" , "C"])

In [7]:
df1

A   B   C
0  10  20  30

# ------Adding more rows to the above data frame.

In [12]:
df1 = pd.DataFrame([[10,20,30] , [70,80,90]] , columns = ["A" , "B" , "C"])

In [13]:
df1

A   B   C
0  10  20  30
1  70  80  90

In [16]:
df1 = pd.DataFrame([[10,20,30] , [70,80,90] , [100,200,300]] , columns = ["A" , "B" , "C"])

In [17]:
df1

A    B    C
0   10   20   30
1   70   80   90
2  100  200  300

# Column-wise.

In [43]:
a = pd.DataFrame([[10,20] , [30,40]] , columns = ["A" , "B"])

In [44]:
a

A   B
0  10  20
1  30  40

In [45]:
b = pd.DataFrame({"A" : [10,30] , "C" :[20,40]})

In [46]:
b

A   C
0  10  20
1  30  40

In [48]:
pd.concat([a,b])

A     B     C
0  10  20.0   NaN
1  30  40.0   NaN
0  10   NaN  20.0
1  30   NaN  40.0

# If we want to identify the rows and columns with respect to the data frames "keys" should be given.

In [57]:
c = pd.concat([a,b] , keys = ["a" , "b"])

# To extract the only a out of the above then we just have to use "c[a]".

In [63]:
c

A     B     C
a 0  10  20.0   NaN
  1  30  40.0   NaN
b 0  10   NaN  20.0
  1  30   NaN  40.0

In [65]:
c.loc["a"]

A     B   C
0  10  20.0 NaN
1  30  40.0 NaN

# To extract the index(0) of the a from above data. Then we can do the foloowing.

In [66]:
c.loc["a" , 0]

A    10.0
B    20.0
C     NaN
Name: (a, 0), dtype: float64

# Note - By default the concat function concatenates the twdo data frames in vertical direction(axis = 0)

# Observations from the above - The a data frame do not have the column "C" so it gave NaN values after concatinating. Likewise the Data frame b has no column "B" so it gave NaN values after concatinating.

# In the above data frame after concartinating, there are duplicate values in the data. So to overcome this problem "ignore_index = True" have to be used.

In [51]:
pd.concat([a,b] , ignore_index = True) # Duplicate values in the index are removed and index is given.

A     B     C
0  10  20.0   NaN
1  30  40.0   NaN
2  10   NaN  20.0
3  30   NaN  40.0

# If axis = 1 (Horizontal operation) is performed on the above to concatenate two data framesa and b.

In [54]:
pd.concat([a,b] , axis = 1)

A   B   A   C
0  10  20  10  20
1  30  40  30  40

# Inner joins and Outer joins.

# ----- If we have a data set A = [1,2,3,4] and data set B = [3,5,6,7] and we combine both of them, Then 4 we have cominations out of it.

# ------------Outer join----------. (A U B)

# Outer join is the one which contains all the values in it including the common value of A and B that is "3".

# Outer join = [1,2,3,4,5,6,7]

# ---------Inner join-----------

# Inner join is the one which contains common values of both A and B which is 3.

# Inner join = [3]

# --------Left Join----------

# left join is the one which contains all the values present towards the left after joining A and B.

# Left join = [1,2,3,4]

# ------Right join-------

# Right join is the one which contains all the values present towards the right after joining A and B.

# Left join = [3,5,6,7]

# Note - While concat function is used, by default it is a outer join as it gives all the values of the data sets.

In [72]:
a = pd.DataFrame([[10,20] , [30,40]] , columns = ["A" , "B"])

In [73]:
a

A   B
0  10  20
1  30  40

In [74]:
b = pd.DataFrame({"A" : [10,30] , "C" :[20,40]})

In [75]:
b

A   C
0  10  20
1  30  40

In [77]:
pd.concat([a,b])

A     B     C
0  10  20.0   NaN
1  30  40.0   NaN
0  10   NaN  20.0
1  30   NaN  40.0

In [78]:
# The above is the outer join.

# Get the inner join of cancatination. 

In [79]:
a

A   B
0  10  20
1  30  40

In [80]:
b

A   C
0  10  20
1  30  40

In [81]:
pd.concat([a,b] , join = "inner")

A
0  10
1  30
0  10
1  30

# In the above as A is common column in bot the data frames a and b. As we used the inner join only A column is returned.

In [83]:
pd.concat([a,b] , axis = 1 , join = "inner")

A   B   A   C
0  10  20  10  20
1  30  40  30  40

# In the above as we are using the axis = 1 condition the operation is performed horixontally. So 10, 20 in the first row, then 30,40 in the second row are common for both the data sets a and b. That is the reason we got all the values of a and b after concatinating.

In [85]:
pd.concat([a,b] , axis = 1 , join = "inner")

A   B   A   C
0  10  20  10  20
1  30  40  30  40

# Also when we observe in the above when both data frames are concatenated, we got bot A columns which are having same values. To overcome this problem we have a method called merge. 

# Concat definition - Concat stacks multiple data frames together along a particular axis.(axis = 0 or axis = 1).

# Merge - Merge combines the data frames based on the values of the shared or common columns.

In [87]:
users = pd.DataFrame({"userid" : [1,2,3] , "name" : ["Sandeep" , "Rishabh" , "Pradeep"]})

In [88]:
users

userid     name
0       1  Sandeep
1       2  Rishabh
2       3  Pradeep

In [95]:
msgs = pd.DataFrame({"userid" : [1,1,2] , "msgs" : ["hmm" , "Acha" , "thik_hai"]})

In [96]:
msgs

userid      msgs
0       1       hmm
1       1      Acha
2       2  thik_hai

# Find the name of a person who sent a particular message. 

In [97]:
users.merge(msgs , on = "userid")

userid     name      msgs
0       1  Sandeep       hmm
1       1  Sandeep      Acha
2       2  Rishabh  thik_hai

# Note - As we can see in the above here, the merging of two data frames by default gives the "inner" join whith respect to the column.

In [109]:
users = pd.DataFrame({"id" : [1,2,3] , "name" : ["Sandeep" , "Rishabh" , "Pradeep"]})

In [110]:
users

id     name
0   1  Sandeep
1   2  Rishabh
2   3  Pradeep

In [111]:
msgs = pd.DataFrame({"userid" : [1,1,2] , "msgs" : ["hmm" , "Acha" , "thik_hai"]})

In [112]:
msgs

userid      msgs
0       1       hmm
1       1      Acha
2       2  thik_hai

# When we have different columns namesbut the values are same in the columns. Then we can merge the data frames in the following method.

In [113]:
users.merge(msgs , left_on = "id" , right_on = "userid")

id     name  userid      msgs
0   1  Sandeep       1       hmm
1   1  Sandeep       1      Acha
2   2  Rishabh       2  thik_hai

# Note - By default when merging "inner" join is done. But the get the "outer" join it should be specifically mentioned as below.

In [115]:
users.merge(msgs , left_on = "id" , right_on = "userid" , how = "outer")

id     name  userid      msgs
0   1  Sandeep     1.0       hmm
1   1  Sandeep     1.0      Acha
2   2  Rishabh     2.0  thik_hai
3   3  Pradeep     NaN       NaN

In [118]:
users.merge(msgs , left_on = "id" , right_on = "userid" , how = "left")

id     name  userid      msgs
0   1  Sandeep     1.0       hmm
1   1  Sandeep     1.0      Acha
2   2  Rishabh     2.0  thik_hai
3   3  Pradeep     NaN       NaN

In [119]:
users.merge(msgs , left_on = "id" , right_on = "userid" , how = "right")

id     name  userid      msgs
0   1  Sandeep       1       hmm
1   1  Sandeep       1      Acha
2   2  Rishabh       2  thik_hai

# Note - So, like abovce when how = "outer" is given, we get all the values of the data frame as it is a outer join.

In [120]:
movies = pd.read_csv(r"C:\Users\DELL\Desktop\Scaler Numpy and Pandas\movies.csv")

In [121]:
movies.head()

Unnamed: 0     id     budget  popularity     revenue  \
0           0  43597  237000000         150  2787965087   
1           1  43598  300000000         139   961000000   
2           2  43599  245000000         107   880674609   
3           3  43600  250000000         112  1084939099   
4           5  43602  258000000         115   890871626   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                              Spider-Man 3           5.9        3576   

   director_id  year month       day  
0         4762  2009   Dec  Thursday  
1         4763  2007   May  Saturday  
2         4764  2015   Oct    Monday  
3         4765  2012   Jul    Monday  
4         4767  2007   May   Tuesday

In [122]:
directors = pd.read_csv(r"C:\Users\DELL\Desktop\Scaler Numpy and Pandas\directors.csv")

In [123]:
directors.head()

Unnamed: 0      director_name    id gender
0           0      James Cameron  4762   Male
1           1     Gore Verbinski  4763   Male
2           2         Sam Mendes  4764   Male
3           3  Christopher Nolan  4765   Male
4           4     Andrew Stanton  4766   Male

# To use the column as an index for the data set index_col = "index of the column" should be given.

In [125]:
movies = pd.read_csv(r"C:\Users\DELL\Desktop\Scaler Numpy and Pandas\movies.csv" , index_col = 0)

In [126]:
movies

id     budget  popularity     revenue  \
0     43597  237000000         150  2787965087   
1     43598  300000000         139   961000000   
2     43599  245000000         107   880674609   
3     43600  250000000         112  1084939099   
5     43602  258000000         115   890871626   
...     ...        ...         ...         ...   
4736  48363          0           3      321952   
4743  48370      27000          19     3151130   
4748  48375          0           7           0   
4749  48376          0           3           0   
4768  48395     220000          14     2040920   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
5                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
4736                            The Last Waltz           7.9          64   
4743                                    Clerks           7.4         755   
4748                                   Rampage           6.0         131   
4749                                   Slacker           6.4          77   
4768                               El Mariachi           6.6         238   

      director_id  year month       day  
0            4762  2009   Dec  Thursday  
1            4763  2007   May  Saturday  
2            4764  2015   Oct    Monday  
3            4765  2012   Jul    Monday  
5            4767  2007   May   Tuesday  
...           ...   ...   ...       ...  
4736         4809  1978   May    Monday  
4743         5369  1994   Sep   Tuesday  
4748         5148  2009   Aug    Friday  
4749         5535  1990   Jul    Friday  
4768         5097  1992   Sep    Friday  

[1465 rows x 11 columns]

In [128]:
directors = pd.read_csv(r"C:\Users\DELL\Desktop\Scaler Numpy and Pandas\directors.csv" , index_col = 0)

In [129]:
directors

director_name    id gender
0         James Cameron  4762   Male
1        Gore Verbinski  4763   Male
2            Sam Mendes  4764   Male
3     Christopher Nolan  4765   Male
4        Andrew Stanton  4766   Male
...                 ...   ...    ...
2344      Shane Carruth  7106   Male
2345   Neill Dela Llana  7107    NaN
2346        Scott Smith  7108    NaN
2347        Daniel Hsia  7109   Male
2348   Brian Herzlinger  7110   Male

[2349 rows x 3 columns]

In [130]:
movies.shape

(1465, 11)

In [132]:
directors.shape

(2349, 3)

In [137]:
movies.head(2)

id     budget  popularity     revenue  \
0  43597  237000000         150  2787965087   
1  43598  300000000         139   961000000   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   

   director_id  year month       day  
0         4762  2009   Dec  Thursday  
1         4763  2007   May  Saturday

In [134]:
directors.head(2)

director_name    id gender
0   James Cameron  4762   Male
1  Gore Verbinski  4763   Male

# In the above data sets the "director_id" column in the movies and "id" column in the directors data set are the same. 

# Now we are merging the data set directors to the movies data set, So that I have information af all the name of directors and their gender for a particular movie.

# First check if all the values of "director_id" column in movies data set are present in "id" column of directors data set. This is to check if there are any differences between the to columns before merging.

In [138]:
movies.head(2)

id     budget  popularity     revenue  \
0  43597  237000000         150  2787965087   
1  43598  300000000         139   961000000   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   

   director_id  year month       day  
0         4762  2009   Dec  Thursday  
1         4763  2007   May  Saturday

In [139]:
directors.head(2)

director_name    id gender
0   James Cameron  4762   Male
1  Gore Verbinski  4763   Male

In [140]:
movies["director_id"].isin(directors["id"])

0       True
1       True
2       True
3       True
5       True
        ... 
4736    True
4743    True
4748    True
4749    True
4768    True
Name: director_id, Length: 1465, dtype: bool

# So in the above output as we can see, the length is shown as 1465, which is the shape of number od rows of movies. So all the values of movies column "director_id" are present in the "id" column of the directors data set.

In [141]:
np.all(movies["director_id"].isin(directors["id"]))

True

# Note - The above is to check if there are all True values for the above statement and it gave a output of True as all the values are True.

In [142]:
movies["director_id"].isin(directors["id"]).value_counts()

True    1465
Name: director_id, dtype: int64

# It can be checked in the above way too.

# Now merging the both Data sets movies and directors.

In [143]:
movies.head(2)

id     budget  popularity     revenue  \
0  43597  237000000         150  2787965087   
1  43598  300000000         139   961000000   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   

   director_id  year month       day  
0         4762  2009   Dec  Thursday  
1         4763  2007   May  Saturday

In [146]:
directors.head(2)

director_name    id gender
0   James Cameron  4762   Male
1  Gore Verbinski  4763   Male

In [147]:
movies.merge(directors , left_on = "director_id" , right_on = "id" , how = "left")

id_x     budget  popularity     revenue  \
0     43597  237000000         150  2787965087   
1     43598  300000000         139   961000000   
2     43599  245000000         107   880674609   
3     43600  250000000         112  1084939099   
4     43602  258000000         115   890871626   
...     ...        ...         ...         ...   
1460  48363          0           3      321952   
1461  48370      27000          19     3151130   
1462  48375          0           7           0   
1463  48376          0           3           0   
1464  48395     220000          14     2040920   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      director_id  year month       day      director_name  id_y gender  
0            4762  2009   Dec  Thursday      James Cameron  4762   Male  
1            4763  2007   May  Saturday     Gore Verbinski  4763   Male  
2            4764  2015   Oct    Monday         Sam Mendes  4764   Male  
3            4765  2012   Jul    Monday  Christopher Nolan  4765   Male  
4            4767  2007   May   Tuesday          Sam Raimi  4767   Male  
...           ...   ...   ...       ...                ...   ...    ...  
1460         4809  1978   May    Monday    Martin Scorsese  4809   Male  
1461         5369  1994   Sep   Tuesday        Kevin Smith  5369   Male  
1462         5148  2009   Aug    Friday           Uwe Boll  5148   Male  
1463         5535  1990   Jul    Friday  Richard Linklater  5535   Male  
1464         5097  1992   Sep    Friday   Robert Rodriguez  5097    NaN  

[1465 rows x 14 columns]

# Note - We have used the how = "left" join because we want all the values of movies data which is on left and also the common values in the directors data set.

# Note - If we can see in the above output the id_y column is created as we already have a "id" column in the movies data set and the old id is given a new name of "id_x".

# Question - Delete the columns "director_id" and "id_y" from the above data as we do not need them.

In [149]:
data = movies.merge(directors , left_on = "director_id" , right_on = "id" , how = "left")

In [152]:
data.head(2)

id_x     budget  popularity     revenue  \
0  43597  237000000         150  2787965087   
1  43598  300000000         139   961000000   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   

   director_id  year month       day   director_name  id_y gender  
0         4762  2009   Dec  Thursday   James Cameron  4762   Male  
1         4763  2007   May  Saturday  Gore Verbinski  4763   Male

In [153]:
data.drop(["director_id" , "id_y"] , axis = 1 , inplace = True)

In [154]:
data

id_x     budget  popularity     revenue  \
0     43597  237000000         150  2787965087   
1     43598  300000000         139   961000000   
2     43599  245000000         107   880674609   
3     43600  250000000         112  1084939099   
4     43602  258000000         115   890871626   
...     ...        ...         ...         ...   
1460  48363          0           3      321952   
1461  48370      27000          19     3151130   
1462  48375          0           7           0   
1463  48376          0           3           0   
1464  48395     220000          14     2040920   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      year month       day      director_name gender  
0     2009   Dec  Thursday      James Cameron   Male  
1     2007   May  Saturday     Gore Verbinski   Male  
2     2015   Oct    Monday         Sam Mendes   Male  
3     2012   Jul    Monday  Christopher Nolan   Male  
4     2007   May   Tuesday          Sam Raimi   Male  
...    ...   ...       ...                ...    ...  
1460  1978   May    Monday    Martin Scorsese   Male  
1461  1994   Sep   Tuesday        Kevin Smith   Male  
1462  2009   Aug    Friday           Uwe Boll   Male  
1463  1990   Jul    Friday  Richard Linklater   Male  
1464  1992   Sep    Friday   Robert Rodriguez    NaN  

[1465 rows x 12 columns]